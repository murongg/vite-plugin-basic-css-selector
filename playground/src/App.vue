<script setup lang="ts">
import { useHead } from '@vueuse/head'

// https://github.com/vueuse/head
// you can use this to manipulate the document head in any components,
// they will be rendered correctly in the html results with vite-ssg
useHead({
  title: 'Playground for Basic CSS Selector',
  meta: [
    { name: 'description', content: 'vite-plugin-basic-css-selector Playground' },
  ],
})
</script>

<template>
  <router-view />
</template>
