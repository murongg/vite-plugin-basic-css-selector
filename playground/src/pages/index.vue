<template>
  <ErrorPopup />
  <div class="grid h-screen w-screen overflow-hidden" style="grid-template-rows: min-content auto;">
    <FilterArea />
    <ResultArea class="flex flex-auto" />
  </div>
</template>
