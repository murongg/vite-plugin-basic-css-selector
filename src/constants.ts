export const HTML_ELEMENT_TAGS = [
  'address', 'article', 'aside', 'footer', 'header', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'hgroup', 'main', 'nav', 'section',
  'blockquote', 'dd', 'div', 'dl', 'dt', 'figcaption', 'figure', 'hr', 'li', 'menu', 'ol', 'p', 'pre', 'ul',
  'a', 'abbr', 'b', 'font-weight', 'bdi', 'bdo', 'br', 'cite', 'code', 'data', 'dfn', 'em', 'i', 'kbd', 'mark', 'q', 'rp', 'ruby', 'rt', 's', 'samp', 'small', 'span', 'strong', 'sub', 'sup', 'time', 'u', 'var', 'wbr',
  'area', 'audio', 'MediaStream', 'img', 'map', 'track', 'video',
  'embed', 'iframe', 'object', 'picture', 'source', 'portal',
  'svg', 'math',
  'canvas',
  'del', 'ins',
  'caption', 'col', 'colgroup', 'table', 'tbody', 'tr', 'td', 'tfoot', 'th', 'thead',
  'button', 'datalist', 'option', 'fieldset', 'label', 'form', 'input', 'legend', 'meter', 'optgroup', 'select', 'output', 'progress', 'textarea',
  'details', 'summary', 'dialog',
  'slot', 'template',
  'acronym', 'bgsound', 'big', 'center', 'content', 'dir', 'font', 'frame', 'frameset', 'image', 'marquee', 'menuitem', 'nobr', 'noembed', 'noframes', 'param', 'plaintext', 'rb', 'rtc', 'shadow', 'strike', 'tt', 'xmp',
]
