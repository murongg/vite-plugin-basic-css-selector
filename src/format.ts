import { css_beautify } from 'js-beautify'

export function cssFormat(code: string, type: 'tab' | 'space' = 'space', indent = 2) {
  return css_beautify(code, {
    indent_size: indent,
    indent_with_tabs: type === 'tab',
  })
}
