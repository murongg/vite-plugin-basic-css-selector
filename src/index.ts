import type { Plugin } from 'vite'
import MagicString from 'magic-string'

function VitePluginBasicCssSelector(basic: string): Plugin {
  return {
    name: 'vite-plugin-basic-css-selector',
    transform(code, id) {
      if (id.endsWith('.css')) {
        const s = new MagicString(code)
        const reg = /(?<=(\.|\#))(.*)(?=\s*\{)/g
        const matches = Array.from(code.matchAll(reg))
        if (!matches.length)
          return

        for (const match of matches) {
          const selector = match[0]
          const dot = match[1]
          const raw = `${dot}${selector}`
          if (raw.includes(basic))
            continue

          const start = match.index! - 1
          const end = start + match[0].length + 1
          const replacement = `${basic} ${raw}`
          s.update(start, end, replacement)
        }
        return {
          code: s.toString(),
          map: s.generateMap(),
        }
      }

      return null
    },
  }
}

export default VitePluginBasicCssSelector
