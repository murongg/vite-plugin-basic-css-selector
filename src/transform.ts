import MagicString from 'magic-string'
import { HTML_ELEMENT_TAGS } from './constants'
import type { Options } from '.'

export function transform(code: string, basic: string, options?: Options) {
  const { enableElementSelector } = options || {}
  const s = new MagicString(code)

  // generate with html element tags reg
  let withHtmlElementTagsReg = ''
  if (enableElementSelector)
    withHtmlElementTagsReg = `|(${HTML_ELEMENT_TAGS.join('|')})`

  const regStr = `((?<=(\\.|\\#))${withHtmlElementTagsReg})(.*)(?=\\s*\\{)`
  const reg = new RegExp(regStr, 'g')
  const matches = Array.from(code.matchAll(reg))

  if (!matches.length)
    return

  for (const match of matches) {
    const selector = match[0]
    const dot = match[2] || ''
    const raw = `${dot}${selector}`
    if (raw.includes(basic))
      continue

    const start = match.index! - 1
    const end = start + selector.length + 1
    const replacement = `${basic} ${raw}`
    s.update(start, end, replacement)
  }
  return {
    code: s.toString(),
    map: s.generateMap(),
  }
}
