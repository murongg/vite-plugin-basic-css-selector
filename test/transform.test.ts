import fs from 'node:fs'
import path from 'node:path'
import { describe, expect, it } from 'vitest'
import { transform } from '../src/transform'

describe('transform', () => {
  it('should basic selector', () => {
    const filePath = path.join(__dirname, './fixtures/simple.css')
    const css = fs.readFileSync(filePath)
    const { code } = transform(css.toString(), '.basic') || {}
    expect(code).toMatchInlineSnapshot(`
      ".basic .logo {
        height: 6em;
        padding: 1.5em;
        will-change: filter;
        transition: filter 300ms;
      }
      .basic .logo:hover {
        filter: drop-shadow(0 0 2em #646cffaa);
      }
      .basic .logo.vanilla:hover {
        filter: drop-shadow(0 0 2em #3178c6aa);
      }

      .basic .card {
        padding: 2em;
      }

      .basic .read-the-docs {
        color: #888;
      }
      "
    `)
  })

  it('should with html element tags', () => {
    const filePath = path.join(__dirname, './fixtures/with-html-element-tags.css')
    const css = fs.readFileSync(filePath)
    const { code } = transform(css.toString(), '.basic', {
      enableElementTag: true,
    }) || {}
    expect(code).toMatchInlineSnapshot(`
      ".basic .logo {
        height: 6em;
        padding: 1.5em;
        will-change: filter;
        transition: filter 300ms;
      }

      .basic .logo:hover{
        filter: drop-shadow(0 0 2em #646cffaa);
      }
      .basic .logo.vanilla:hover {
        filter: drop-shadow(0 0 2em #3178c6aa);
      }

      .basic .card {
        padding: 2em;
      }

      .basic .read-the-docs {
        color: #888;
      }
      .basic button {
        color: #888;
      }

      .basic .button {
        color: #909090;
      }
      .basic a {
        color: #888;
      }
      .basic ul {
        color: #888;
      }

      d {
        color: #888;
      }
      "
    `)
  })
})
