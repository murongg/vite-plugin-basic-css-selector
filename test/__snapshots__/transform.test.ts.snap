// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transform > should basic selector 1`] = `
".basic .logo {
  height: 6em;
  padding: 1.5em;
  will-change: filter;
  transition: filter 300ms;
}

.basic .logo:hover {
  filter: drop-shadow(0 0 2em #646cffaa);
}

.basic .logo.vanilla:hover {
  filter: drop-shadow(0 0 2em #3178c6aa);
}

.basic .card {
  padding: 2em;
}

.basic .read-the-docs {
  color: #888;
}

* {
  font-size: 12px;
}

*.work {
  font-family: 'Franklin Gothic Medium', 'Arial Narrow', Arial, sans-serif;
}

*#id {
  font-family: 'Franklin Gothic Medium', 'Arial Narrow', Arial, sans-serif;
}

a_button_a {
  color: #888;
}

.basic .button {
  color: #909090;
}

button {
  color: #888;
}
"
`;

exports[`transform > should full case 1`] = `
".basic * {
  font-size: 12px;
}

.basic *.work {
  font-family: 'Franklin Gothic Medium', 'Arial Narrow', Arial, sans-serif;
}

.basic *#id {
  font-family: 'Franklin Gothic Medium', 'Arial Narrow', Arial, sans-serif;
}

.basic .read-the-docs {
  color: #888;
}

.basic .logo:hover {
  filter: drop-shadow(0 0 2em #646cffaa);
}

a_button_a {
  color: #888;
}

.basic .button {
  color: #909090;
}
.basic button {
  color: #888;
}
"
`;

exports[`transform > should universal selectors 1`] = `
".basic * {
  font-size: 12px;
}

.basic *.work {
  font-family: 'Franklin Gothic Medium', 'Arial Narrow', Arial, sans-serif;
}

.basic *#id {
  font-family: 'Franklin Gothic Medium', 'Arial Narrow', Arial, sans-serif;
}
"
`;

exports[`transform > should with html element tags 1`] = `
".basic .logo {
  height: 6em;
  padding: 1.5em;
  will-change: filter;
  transition: filter 300ms;
}

.basic .logo:hover {
  filter: drop-shadow(0 0 2em #646cffaa);
}

.basic .logo.vanilla:hover {
  filter: drop-shadow(0 0 2em #3178c6aa);
}

.basic .card {
  padding: 2em;
}

.basic .read-the-docs {
  color: #888;
}
.basic button {
  color: #888;
}

.basic .button {
  color: #909090;
}
.basic a {
  color: #888;
}
.basic ul {
  color: #888;
}

d {
  color: #888;
}
"
`;

exports[`transform > should with only html element tags 1`] = `
".basic button {
  color: #888;
}

button_a {
  color: #888;
}

a_button {
  color: #888;
}

a_button_a {
  color: #888;
}

.basic .button {
  color: #909090;
}
.basic a {
  color: #888;
}
.basic ul {
  color: #888;
}

d {
  color: #888;
}
"
`;
